### ISSUE TITLE
Error 500 when trying to use GitHub authentication

### ISSUE THREAD

**Issue by mcanevet** [go-gitea/gitea#3652]
- Gitea version (or commit ref): 1.4.0+rc2-2-g92a3061
- Git version:
- Operating system:
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist: https://gist.github.com/mcanevet/5e04211b45861ad94557badc3b774654

## Description
I configured an oauth2 authentication pointing to GitHub.
When I click on "Sign with GitHub", I get an Error 500 with the message "An error has occurred : remove /app/gitea/data/sessions/oauth2/session_: no such file or directory".
My Gitea is behind an HTTP proxy so maybe it's the origin of the problem, but I'm not sure.

**Comment by mmarif4u** [comment]
Similar #3644?

**Comment by lafriks** [comment]
Seems to be related to markbates/goth#192

**Comment by Treora** [comment]
Experiencing a similar/same problem: clicking "Sign in with" for either GitHub/GitLab (haven't tried others) leads to a 500 error. No error message for me, just this (for GitLab, which has authentication name 'gitlab'):

    server_1  | [Macaron] 2018-03-09 10:34:52: Completed GET /user/oauth2/gitlab 404 Not Found in 25.06562ms

(note the 404 in the logs, resulting in a 500 page somehow)

PS: Gitea version is 1a83581, from the `gitea/gitea:latest` docker image.

**Comment by Treora** [comment]
Also, possibly related(?), entering a non-existent name results in an internal server error:

```
server_1  | [Macaron] 2018-03-09 10:58:21: Started GET /user/oauth2/lalala for 172.20.0.1
server_1  | [Macaron] PANIC: runtime error: invalid memory address or nil pointer dereference
```

<details><summary>Full error log</summary>

```
server_1  | [Macaron] 2018-03-09 10:58:21: Started GET /user/oauth2/lalala for 172.20.0.1
server_1  | [Macaron] PANIC: runtime error: invalid memory address or nil pointer dereference
server_1  | /usr/lib/go/src/runtime/panic.go:491 (0x5615968c0a56)
server_1  | /usr/lib/go/src/runtime/panic.go:63 (0x5615968bf92f)
server_1  | /usr/lib/go/src/runtime/signal_unix.go:367 (0x5615968d827f)
server_1  | /srv/app/src/code.gitea.io/gitea/routers/user/auth.go:462 (0x56159734dabd)
server_1  | /srv/app/src/code.gitea.io/gitea/routers/user/auth.go:374 (0x56159734cc7f)
server_1  | /usr/lib/go/src/runtime/asm_amd64.s:509 (0x5615968f040c)
server_1  | /usr/lib/go/src/reflect/value.go:434 (0x56159695a016)
server_1  | /usr/lib/go/src/reflect/value.go:302 (0x5615969595d5)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:177 (0x561596c6178a)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:137 (0x561596c6107b)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:121 (0x561596c9064f)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:112 (0x561596c90587)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/github.com/go-macaron/session/session.go:186 (0x561596ce7698)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:79 (0x561596c90432)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:157 (0x561596c61445)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:135 (0x561596c6116c)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:121 (0x561596c9064f)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:112 (0x561596c90587)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/recovery.go:161 (0x561596ca286c)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/logger.go:40 (0x561596c9397d)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:157 (0x561596c61445)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:135 (0x561596c6116c)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:121 (0x561596c9064f)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:112 (0x561596c90587)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/logger.go:52 (0x561596ca1afd)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/logger.go:40 (0x561596c9397d)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:157 (0x561596c61445)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/github.com/go-macaron/inject/inject.go:135 (0x561596c6116c)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/context.go:121 (0x561596c9064f)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/router.go:187 (0x561596ca3b54)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/router.go:303 (0x561596c9d03d)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/macaron.go:220 (0x561596c94f09)
server_1  | /srv/app/src/code.gitea.io/gitea/vendor/github.com/gorilla/context/context.go:141 (0x561596f825dc)
server_1  | /usr/lib/go/src/net/http/server.go:1918 (0x561596b2f295)
server_1  | /usr/lib/go/src/net/http/server.go:2619 (0x561596b31fb5)
server_1  | /usr/lib/go/src/net/http/server.go:1801 (0x561596b2e14e)
server_1  | /usr/lib/go/src/runtime/asm_amd64.s:2337 (0x5615968f2cb0)
server_1  | [Macaron] 2018-03-09 10:58:21: Completed GET /user/oauth2/lalala 500 Internal Server Error in 7.004216ms
```
</details>

**Comment by lafriks** [comment]
I have provided fix to upstream library markbates/goth#217
